client/v2/tx/factory.go

```diff
@@ -0,0 +1,588 @@
+
+func (f Factory) AccountNumber() uint64              { return f.accountNumber }
+func (f Factory) Sequence() uint64                   { return f.sequence }
+func (f Factory) Gas() uint64                        { return f.gas }
+func (f Factory) GasAdjustment() float64             { return f.gasAdjustment }
+func (f Factory) Keybase() keyring.Keyring           { return f.keybase }
+func (f Factory) ChainID() string                    { return f.chainID }
+func (f Factory) Memo() string                       { return f.memo }
+func (f Factory) Fees() sdk.Coins                    { return f.fees }
+func (f Factory) GasPrices() sdk.DecCoins            { return f.gasPrices }
+func (f Factory) AccountRetriever() AccountRetriever { return f.accountRetriever }
+func (f Factory) TimeoutHeight() uint64              { return f.timeoutHeight }
+func (f Factory) FromName() string                   { return f.fromName }
+func (f Factory) SimulateAndExecute() bool           { return f.simulateAndExecute }
+func (f Factory) SignMode() signing.SignMode         { return f.signMode }
+
+// getSimPK gets the public key to use for building a simulation tx.
+// Note, we should only check for keys in the keybase if we are in simulate and execute mode,
+// e.g. when using --gas=auto.
+// When using --dry-run, we are is simulation mode only and should not check the keybase.
+// Ref: https://github.com/cosmos/cosmos-sdk/issues/11283
+func (f Factory) getSimPK() (cryptotypes.PubKey, error) {
+	var (
+		err error
+		pk  cryptotypes.PubKey = &secp256k1.PubKey{} // use default public key type
+	)
+
+	if f.simulateAndExecute && f.keybase != nil {
+		pk, err = f.keybase.GetPubKey(f.fromName)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return pk, nil
+}
+
+// getSimSignatureData based on the pubKey type gets the correct SignatureData type
+// to use for building a simulation tx.
+func (f Factory) getSimSignatureData(pk cryptotypes.PubKey) signing.SignatureData {
+	multisigPubKey, ok := pk.(*multisig.LegacyAminoPubKey) // TODO: abstract out multisig pubkey
+	if !ok {
+		return &signing.SingleSignatureData{SignMode: f.signMode}
+	}
+
+	multiSignatureData := make([]signing.SignatureData, 0, multisigPubKey.Threshold)
+	for i := uint32(0); i < multisigPubKey.Threshold; i++ {
+		multiSignatureData = append(multiSignatureData, &signing.SingleSignatureData{
+			SignMode: f.SignMode(),
+		})
+	}
+
+	return &signing.MultiSignatureData{
+		Signatures: multiSignatureData,
+	}
+}
+
+// Sign signs a given tx with a named key. The bytes signed over are canconical.
+// The resulting signature will be added to the transaction builder overwriting the previous
+// ones if overwrite=true (otherwise, the signature will be appended).
+// Signing a transaction with mutltiple signers in the DIRECT mode is not supported and will
+// return an error.
+// An error is returned upon failure.
+func Sign(ctx context.Context, txf Factory, name string, txBuilder TxBuilder, overwriteSig bool) error {
```

Comment:
Moved the `Sign` func from tx/tx.go to here (is not working yet, though), `Factory` should be able to build both Unsigned and Signed txs.